lib/proxy/client.go

```diff
@@ -0,0 +1,558 @@
+	for _, id := range proxyIDs {
+		ids[id] = struct{}{}
+
+		conn, ok := c.conns[id]
+		if !ok {
+			c.metrics.reportTunnelError(errorExistingTunnelNotFound)
+			continue
+		}
+
+		var err error
+		stream, err = c.startStream(conn)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		break
+	}
+	c.RUnlock()
+
+	if stream != nil {
+		return stream, true, nil
+	}
+
+	if len(errs) != 0 {
+		c.metrics.reportTunnelError(errorExistingTunnelInvalid)
+		return nil, false, trace.NewAggregate(errs...)
+	}
+
+	// proxy does not exist in cache.
+	// get list of proxies directly from auth.
+	c.Lock()
+	defer c.Unlock()
+	proxies, err := c.config.AuthClient.GetProxies()
+	if err != nil {
+		c.metrics.reportTunnelError(errorAuthClient)
+		return nil, false, trace.NewAggregate(errs...)
+	}
+
+	errs = nil
+	for _, proxy := range proxies {
+		id := proxy.GetName()
+		if _, ok := ids[id]; !ok {
+			continue
+		}
+
+		conn, err := c.connect(id, proxy.GetPeerAddr())
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		stream, err := c.startStream(conn)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		c.conns[id] = conn
+		return stream, false, nil
+	}
+
+	c.metrics.reportTunnelError(errorProxiesUnavailable)
+	return nil, false, trace.NotFound("Error dialling all proxies: %+v", trace.NewAggregate(errs...).Error())
+}
+
+// connect dials a new connection to proxyAddr.
+func (c *Client) connect(id string, proxyPeerAddr string) (*clientConn, error) {
+	tlsConfig, err := c.config.getConfigForServer()
+	if err != nil {
+		c.metrics.reportTunnelError(errorNewTunnelDial)
+		return nil, trace.Wrap(err, "Error updating client tls config")
+	}
+
+	connCtx, cancel := context.WithCancel(c.ctx)
+	wg := new(sync.WaitGroup)
+
+	transportCreds := newProxyCredentials(credentials.NewTLS(tlsConfig))
+	conn, err := grpc.DialContext(
+		connCtx,
+		proxyPeerAddr,
+		grpc.WithTransportCredentials(transportCreds),
+		grpc.WithStatsHandler(newStatsHandler(c.metrics)),
+		grpc.WithChainStreamInterceptor(metadata.StreamClientInterceptor, utils.GRPCClientStreamErrorInterceptor, streamCounterInterceptor(wg)),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                peerKeepAlive,
+			Timeout:             peerTimeout,
+			PermitWithoutStream: true,
+		}),
+	)
+	if err != nil {
+		c.metrics.reportTunnelError(errorNewTunnelDial)
+		return nil, trace.Wrap(err, "Error dialling proxy %+v", id)
+	}
+
+	return &clientConn{
+		ClientConn: conn,
+		ctx:        connCtx,
+		cancel:     cancel,
+		wg:         wg,
+		id:         id,
+		addr:       proxyPeerAddr,
+	}, nil
+}
+
+// startStream opens a new stream to the provided connection.
+func (c *Client) startStream(conn *clientConn) (clientapi.ProxyService_DialNodeClient, error) {
+	client := clientapi.NewProxyServiceClient(conn.ClientConn)
+
+	stream, err := client.DialNode(conn.ctx)
+	if err != nil {
+		c.metrics.reportTunnelError(errorStreamStart)
```

Comment:
Should this also check the connectivity state and report an `errorTransientFailure` like in `testConnection`? Or is that error only meant to be recorded during the connection tests? If it's not just meant for `testConnection` there might be more places than this which should be checking the connectivity state

Reply:
I think the logic is fine - I was more just questioning if the tunnel error being reported should be more reflective here. 
```suggestion
	      if conn.GetState() == connectivity.TransientFailure {
			c.metrics.reportTunnelError(errorTransientFailure)
		} else {
		        c.metrics.reportTunnelError(errorStreamStart)
		}
```

I guess what I'm trying to ask is would being more explicit here about what gets reported aid in debugging/diagnosing issues? Although I suppose if there is an error returned by `DialNode` there is a pretty good chance the connectivity state is `TransientFailure`

Also now that I think about it more should `errorTransientFailure` be something more descriptive and less grpc-like. Maybe something like `errorConnectionFailure = "PEER_CONNECTION_FAILURE` ?